exec/cnex/cell.c

```diff
@@ -0,0 +1,406 @@
+        if (c->array == NULL) {
+            fatal_error("Unable to reallcoate memory for write array.");
+        }
+        for (n = c->array_size; n < i; n++) {
+            cell_resetCell(&c->array[n]);
+        }
+        c->array_size = i+1;
+    }
+    return &c->array[i];
+}
+
+Cell *cell_createDictionaryCell(size_t iEntries)
+{
+    Cell *c = cell_newCell();
+
+    c->type = cDictionary;
+
+    /* ToDo: Implement dictionary Cell Types */
+    assert(c->type != cDictionary);
+
+    return c;
+}
+
+Cell *cell_createStringCell(size_t length)
+{
+    Cell *c = cell_newCell();
+
+    c->type = cString;
+    c->string = string_newString();
+    c->string->length = length;
+    c->string->data = malloc(length);
+    if (!c->string->data) {
+        fatal_error("Unable to allocate memory for requested string with length of %ld.", length);
+    }
+    return c;
+}
+
+void cell_copyCell(Cell *dest, const Cell *source)
+{
+    assert(source != NULL);
+    assert(dest != NULL);
+
+    cell_clearCell(dest);
+
+    dest->number = source->number;
+    if (source->type == cString && source->string != NULL) {
+        dest->string = string_copyString(source->string);
+    } else {
+        dest->string = NULL;
+    }
+    dest->array_size = source->array_size;
+    if (source->type == cArray && source->array != NULL) {
+        uint32_t i = 0;
+        dest->array = malloc(sizeof(Cell) * dest->array_size);
+        for (i = 0; i < dest->array_size; i++) {
+            cell_copyCell(&dest->array[i], &source->array[i]);
+        }
+    } else {
+        dest->array = NULL;
+    }
+    dest->address = source->address;
+    dest->boolean = source->boolean;
+    dest->type = source->type;
+}
+
+int32_t cell_compareCell(const Cell * s, const Cell * d)
+{
+    assert(s->type == d->type);
+    /* ToDo: Fix this, so that it actually COMPARES two cells, and returns 0 if equal, -1 if less than, and > 0 if greater than. */
+    if (s->type != d->type) {
+        return -1;
+    }
+
+    switch (s->type) {
+        case cString:   return string_compareString(s->string, d->string);
+        case cAddress:  return s->address == d->address != 0;
+        case cBoolean:  return s->boolean == d->boolean != 0;
+        case cArray:    return s->array_size == d->array_size != 0;
+        case cNumber:   return bid128_quiet_equal(s->number, d->number) != 0;
+    }
+    return 1;
+}
+
+Cell *cell_newCell()
+{
+    Cell *c = malloc(sizeof(Cell));
+    if (c == NULL) {
+        fatal_error("Could not allocate new cell object.");
+    }
+
+    c->number = bid128_from_uint32(0);
+    c->string = NULL;
+    c->address = NULL;
+    c->boolean = FALSE;
+    c->array = NULL;
+    c->array_size = 0;
+    c->type = cNothing;
+    return c;
+}
+
+Cell *cell_newCellType(CellType t)
+{
+    Cell *c = cell_newCell();
+    c->type = t;
+
+    return c;
+}
+
+void cell_resetCell(Cell *c)
+{
+    c->number = bid128_from_uint32(0);
+    c->string = NULL;
```

Comment:
No, because resetCell() is called from the RESETC opcode.  Perhaps, resetCell() should in fact free() any objects that might exist when it's called.  Originally, there was nothing to free() on a reset.  Now, there is.